src/zulip/api.rs

```diff
@@ -144,6 +174,75 @@ impl ZulipApi {
         Ok(())
     }
 
+    pub(crate) fn update_stream_membership(
+        &self,
+        stream_name: &str,
+        stream_id: u64,
+        add_ids: &[u64],
+        remove_ids: &[u64],
+    ) -> anyhow::Result<()> {
+        if add_ids.is_empty() && remove_ids.is_empty() {
+            log::debug!(
+                "stream {} does not need to have its members updated",
+                stream_id
+            );
+            return Ok(());
+        }
+
+        log::info!(
+            "updating stream {} by adding {:?} and removing {:?}",
+            stream_id,
+            add_ids,
+            remove_ids
+        );
+
+        if self.dry_run {
+            return Ok(());
+        }
+
+        let subscriptions = format!("{{\"name\": \"{stream_name}\", \"description\": \"\"}}");
+        let add_ids = serialize_as_array(add_ids);
+        let mut form = HashMap::new();
+        form.insert("subscriptions", subscriptions.as_str());
+        form.insert("principals", add_ids.as_str());
+
+        let path = format!("/users/me/subscriptions");
+        let response = self.req(reqwest::Method::POST, &path, Some(form))?;
+
+        if response.status() == 400 {
+            log::warn!(
+                "failed to update group membership with a bad request: {}",
+                response
+                    .text()
+                    .unwrap_or_else(|_| String::from("<BODY NOT DECODABLE>"))
+            );
+            return Ok(());
+        }
+        response.error_for_status()?;
+
+        let subscriptions = format!("[{stream_name}]");
+        let remove_ids = serialize_as_array(remove_ids);
+        let mut form = HashMap::new();
+        form.insert("subscriptions", subscriptions.as_str());
+        form.insert("principals", remove_ids.as_str());
+
+        let path = format!("/users/me/subscriptions");
+        let response = self.req(reqwest::Method::DELETE, &path, Some(form))?;
+
+        if response.status() == 400 {
+            log::warn!(
+                "failed to update stream membership with a bad request: {}",
+                response
+                    .text()
+                    .unwrap_or_else(|_| String::from("<BODY NOT DECODABLE>"))
+            );
+            return Ok(());
+        }
+        response.error_for_status()?;
```

Comment:
this part here is very similar with above. Can you extract it in some way to remove code duplication?

Reply:
Fixed!